lib/awful/hotkeys_popup/widget.lua

```diff
@@ -141,53 +141,79 @@ function widget.new(args)
-            ['#73']="F7",
-            ['#74']="F8",
-            ['#75']="F9",
-            ['#76']="F10",
-            ['#95']="F11",
-            ['#96']="F12",
-            ['#10']="1",
-            ['#11']="2",
-            ['#12']="3",
-            ['#13']="4",
-            ['#14']="5",
-            ['#15']="6",
-            ['#16']="7",
-            ['#17']="8",
-            ['#18']="9",
-            ['#19']="0",
-            ['#20']="-",
-            ['#21']="=",
-            Control="Ctrl"
+            Mod4       ="Super",
+            Mod1       ="Alt",
+            Escape     ="Esc",
+            Insert     ="Ins",
+            Delete     ="Del",
+            Backspace  ="BackSpc",
+            Return     ="Enter",
+            Next       ="PgDn",
+            Prior      ="PgUp",
+            ['#108']   ="Alt Gr",
+            Left       ="←",
+            Up         ="↑",
+            Right      ="→",
+            Down       ="↓",
+            ['#67']    ="F1",
+            ['#68']    ="F2",
+            ['#69']    ="F3",
+            ['#70']    ="F4",
+            ['#71']    ="F5",
+            ['#72']    ="F6",
+            ['#73']    ="F7",
+            ['#74']    ="F8",
+            ['#75']    ="F9",
+            ['#76']    ="F10",
+            ['#95']    ="F11",
+            ['#96']    ="F12",
+            ['#10']    ="1",
+            ['#11']    ="2",
+            ['#12']    ="3",
+            ['#13']    ="4",
+            ['#14']    ="5",
+            ['#15']    ="6",
+            ['#16']    ="7",
+            ['#17']    ="8",
+            ['#18']    ="9",
+            ['#19']    ="0",
+            ['#20']    ="-",
+            ['#21']    ="=",
+            [' ']      ="space",
+            Control    ="Ctrl",
+            KP_End     ="Num1",
+            KP_Down    ="Num2",
+            KP_Next    ="Num3",
+            KP_Left    ="Num4",
+            KP_Begin   ="Num4",
+            KP_Right   ="Num6",
+            KP_Home    ="Num7",
+            KP_Up      ="Num8",
+            KP_Prior   ="Num9",
+            KP_Insert  ="Num0",
+            KP_Delete  ="Num.",
+            KP_Divide  ="Num⁄", -- a fraction slash, to avoid confusion with
+                                -- the solidus, which separates keys.
+            KP_Multiply="Num*",
+            KP_Subtract="Num-",
+            KP_Add     ="Num+",
+            KP_Enter   ="NumEnter",
         },
         _additional_hotkeys = {},
         _cached_wiboxes = {},
         _cached_awful_keys = {},
         _colors_counter = {},
         _group_list = {},
         _widget_settings_loaded = false,
+        _keygroups = {},
     }
+    for k, v in pairs(awful.key.keygroups) do
+        widget_instance._keygroups[k] = {}
+        for k2, v2 in pairs(v) do
+            widget_instance._keygroups[k][k2] = widget_instance.labels[v2[1]] or v2[1]
+        end
+        widget_instance._keygroups[k]["text"] = table.concat(widget_instance._keygroups[k], "/")
```

Comment:
why you storing keys as concatted with '/' instead of storing them as a table?

that would avoid false-positive matching in `first = string.find(v.text, value.key, 1, true)` next in your code

Reply:
It's precisely because I found a text match easier to program than finding every key in a table.

There should be no real chance of false-positives. Abbreviations only trigger with five or more keys, `/` is only ever used to separate keys(*), and it doesn't matter if two keygroups share keys because it doesn't check for the name of the keygroup.

In your previous example, if `numrowA="1/2/3/4/5/6/7/8"` and `numrowB="2/3/4/5/6/7/8/9"`, the abbreviation's still going to be `2…6` regardless of which keygroup is tested first.

There _is_ a hypothetical false-positive, now you mention it, but this would only happen if the first/last key in the cheatsheet matches part of the name in a keygroup. For example, testing `2/3/4/5/6` against `Num2/3/4/5/6/7`. It still can be worked around with a regexp.

(*) I took care of defining the `/` key in the alias list as `Num⁄`, which doesn't use an actual `/` but a different Unicode character. I'm not sure if any keyboard has a dedicated `/` outside the numeric keypad (mine doesn't), but if that's the case it can be added to the aliases list, too.

Reply:
example of false positive is `KP_Enter/...` will match `n`, `t`, `e`, etc

it must be a table

Reply:
That's the reason of the minimum. This method works `after` keys with the same description are merged, and counts the keys in the merged entries before trying to abbreviate anything.

In an entry like `Super+n/t`, `n/t` only contains two keys, so the subroutine aborts before trying to find a match. The same for `t`, `n`, `t/n/e` or even `n/t/e/r`, because the minimum is five.

And `n/t/e/r` isn't a subtring of `Enter`, so there's no match anyways.

Reply:
> It still can be worked around with a regexp.

that still enough of reason to refactor it to be a table

Reply:
I uploaded a patch that uses no regexps and no tables.

By the way, how can I use your static code check tools locally? I couldn't find at `CONTRIBUTING.md` how to replicate the tests done with Travis CI.

Reply:
you can look for it in travis's config file accordingly

in this case it's `make check`